MovieTime/MovieTime/Service/MTParameterEncoding.swift

```diff
@@ -0,0 +1,375 @@
+            return urlRequest
+        }
+
+        do {
+            let data = try JSONSerialization.data(withJSONObject: parameters, options: options)
+
+            if urlRequest.value(forHTTPHeaderField: "Content-Type") == nil {
+                urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
+            }
+
+            urlRequest.httpBody = data
+        } catch {
+            return urlRequest
+        }
+
+        return urlRequest
+    }
+}
+
+// MARK: - ArrayEncoding
+private let MTArrayParametersKey = "MTArrayParametersKey"
+
+public struct ArrayEncoding: MTParameterEncoding {
+
+    // MARK: Properties
+
+    /// Returns a `ArrayEncoding` instance with default writing options.
+    public static var `default`: ArrayEncoding { return ArrayEncoding() }
+
+    public static var prettyPrinted: ArrayEncoding { return ArrayEncoding(options: .prettyPrinted) }
+
+    public static var customQueryString: ArrayEncoding { return ArrayEncoding(customize: .reflection) }
+
+    public let options: JSONSerialization.WritingOptions
+    public let customize: Customize
+
+    public init(customize: Customize = .none, options: JSONSerialization.WritingOptions = []) {
+        self.options = options
+        self.customize = customize
+    }
+
+    public func urlEncode(_ url: String, method: MTHTTPMethod, with parameters: MTParameters?) -> URL? {
+        var url = url
+        if customize == .reflection {
+            // Encodes parameters in URL
+            if method.encodesParametersInURL() {
+                url = url.reflectionEncode(parameters: parameters)
+            }
+        }
+        guard let encodedURL = url.addingPercentEncoding(withAllowedCharacters: NSCharacterSet.urlQueryAllowed) else { return nil }
+
+        return URL(string: encodedURL)
+    }
+
+    public func encode(_ urlRequest: URLRequest, with parameters: MTParameters?) -> URLRequest {
+        var urlRequest = urlRequest
+
+        guard let parameters = parameters, let array = parameters[MTArrayParametersKey] else {
+            return urlRequest
+        }
+        guard let method = MTHTTPMethod(rawValue: urlRequest.httpMethod ?? "GET"), !method.encodesParametersInURL() else {
+            return urlRequest
+        }
+
+        do {
+            let data = try JSONSerialization.data(withJSONObject: array, options: options)
+
+            if urlRequest.value(forHTTPHeaderField: "Content-Type") == nil {
+                urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
+            }
+
+            urlRequest.httpBody = data
+        } catch {
+            return urlRequest
+        }
+
+        return urlRequest
+    }
+}
+
+// MARK: - Custom Extensions
+
+/// Extension that allows an array be sent as a request parameters
+extension Array {
+    func asParameters() -> MTParameters {
+        return [MTArrayParametersKey: self]
+    }
+}
+
+extension NSArray {
```

Comment:
when do you need to use NSArray in this app?

Reply:
I did not think about it during the test, but I know that if I need to have an array that holds any Objective-C object, I use an NSArray.
Also, Array is a value type and NSArray is a reference type, so if I want to compare instance with `===`, or if I want to create a mutable state class, I could use NSArray.

Reply:
For the most part, Swift arrays can be used in place of NSArrays. A reason to use NSArrays are when they have some method/property I need.